packages/amino/src/signdoc.ts

```diff
@@ -69,5 +69,51 @@ export function makeSignDoc(
 }
 
 export function serializeSignDoc(signDoc: StdSignDoc): Uint8Array {
-  return toUtf8(sortedJsonStringify(signDoc));
-}
+  const fixedMsgs = signDoc.msgs.map((aminoMsg) => {
+    if (aminoMsg.type.startsWith("regen")) {
+      // regen ledger incorrectly implements amino signing.
+      // this short term workaround makes cosmjs compatible with regen-ledger v4.0
+      // until regen ledger upgrades with the fix in v5.0
+      // ref: https://github.com/regen-network/regen-ledger/pull/1480
+
+      console.log("unstripped item:", aminoMsg.value)
+      var foo = stripTypeFields(aminoMsg.value)
+      console.log("stripped item:", foo)
+      return foo
+    } else {
+      return aminoMsg
+    }
+  })
+
+  function stripTypeFields(msg: any): any {
+    var res: any = {};
+    for (const [key, val] of Object.entries(msg)) {
+      // if the key is '$type' we don't include it in the result
+      // relace below with "if (key != '$type' && val != ''){" to get regen-js test passing
+      if (key != '$type'){
+        var strippedValue: any;
+        if (Array.isArray(val)){
+          strippedValue = [];
+          for(const elt of val) {
+            if(typeof elt === 'object'){
+              strippedValue.push(stripTypeFields(elt))
+            } else {
+              strippedValue.push(elt)
+            }
+          }
+        } else if (typeof val === 'object'){
+          strippedValue = stripTypeFields(val)
+        } else {
+          strippedValue = val
+        }
+        
+        res[key] = strippedValue;
+      }
+    }
+    return res
+  }
```

Comment:
I think we can leave this out. We are able to control the shape of the AminoTypes in regen-js at this level, which means we can manually omit the nested `$type`s and also make sure empty strings are converted to undefined.

If we use this stripping method, I feel there is a risk of hard-to-predict side effects.

```suggestion
```

Reply:
See https://github.com/regen-network/regen-js/pull/38/files#diff-26b2b3ecc876221c1986b1046cd3f7164c19ff3bdf6e4bd12e8be3a0f29b7b64 for example

Reply:
Cool - I just removed this!